### ISSUE TITLE
Implement StateFraudProof

### ISSUE THREAD

**Issue by nusret1996** [celestiaorg/celestia-core#480]
### Summary
We need to provide support for 'StateFraudProof's, which indicate to the light clients that there is an invalid state within a finalized block.

### Details
'StateFraudProof's show that an intermediate state root attested by the [availableDataRoot](https://github.com/celestiaorg/celestia-specs/blob/master/src/specs/data_structures.md#header) of a block is invalid given the witness state elements and the transactions. Like 'BadEncodingFraudProof's, 'StateFraudProof's will almost never be generated, as a valid proof means that at least 2/3 of the validators have signed a block with an invalid state. In this context, existence of 'StateFraudProof's provide a game theoretic guarantee of validity for the states within the blocks.  

Data structure for the StateFraudProof can be found [here](https://github.com/celestiaorg/celestia-specs/blob/d56976387039a1fb319b5397afca19e288daa583/src/specs/networking.md#statefraudproof). We assume a deferred execution model for the states, thus, intermediate state roots and transaction data will be in different blocks at consecutive heights.

### Action Items
- [ ] Write the pseudocode for the methods to create and verify 'StateFraudProof's: VerifyStateFraudProof() and CreateStateFraudProof()
- [ ] Codify the pseudocodes for the methods above.
- [ ] Add descriptions of the code to the spec.

### References
Tracking issue #459 
Blocked by #460, #462

**Comment by evan-forbes** [comment]
important notes in [#132](https://github.com/celestiaorg/optimint/issues/132)

**Comment by liamsi** [comment]
Let's focus on fraud proofs for Rollups only instead. Let's revisit this whenever we are ready to have fraud proofs for the main chain too.
